trace/server/src/main/scala/org/http4s/otel4s/middleware/trace/server/ServerMiddleware.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2023 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+package otel4s.middleware
+package trace
+package server
+
+import cats.data.Kleisli
+import cats.effect.kernel.MonadCancelThrow
+import cats.effect.kernel.Outcome
+import cats.effect.syntax.monadCancel._
+import cats.syntax.applicative._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import fs2.Stream
+import org.typelevel.otel4s.KindTransformer
+import org.typelevel.otel4s.trace.SpanKind
+import org.typelevel.otel4s.trace.StatusCode
+import org.typelevel.otel4s.trace.Tracer
+import org.typelevel.otel4s.trace.TracerProvider
+
+/** Middleware for wrapping [[org.http4s.HttpApp `HttpApp`]],
+  * [[org.http4s.HttpRoutes `HttpRoutes`]], and arbitrary
+  * [[org.http4s.Http `Http`]] instances to add tracing.
+  *
+  * @see [[https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-server]]
+  */
+sealed trait ServerMiddleware[F[_]] {
+
+  /** Wraps an [[org.http4s.Http `Http`]] to add tracing in a way that
+    * abstracts over [[org.http4s.HttpApp `HttpApp`]] and
+    * [[org.http4s.HttpRoutes `HttpRoutes`]]. In most cases, it is preferable
+    * to use the methods that directly wrap the specific desired type.
+    *
+    * @return a wrapped `Http` with added tracing
+    * @see [[wrapHttpApp]]
+    * @see [[wrapHttpRoutes]]
+    */
+  def wrapGenericHttp[G[_]: MonadCancelThrow](
+      f: Http[G, F]
+  )(implicit kt: KindTransformer[F, G]): Http[G, F]
+
+  /** @return a wrapped [[org.http4s.HttpApp `HttpApp`]] with added tracing */
+  def wrapHttpApp(httpApp: HttpApp[F]): HttpApp[F]
```

Comment:
this can trivially be used with `.withHttpWebSocketApp` since it is pure rather than effectful